### ISSUE TITLE
Intermittent alsa buffer underrun

### ISSUE THREAD

**Issue by koltenpearson** [libsdl-org/SDL#9171]
Using SDL 2.30.0, I am occasionally getting a buffer underrun when attempting to use SDL_QueueAudio. Changing the buffer size or other SDL_AudioSpec options does not seem to make a difference in its rate of occurrence. It happens at the start of the program and persists throughout its execution, or not at all.

I was able to boil it down to a reproducible snippet
```
#include <SDL2/SDL.h>

int main(int argc, char** argv) {
    SDL_version v;
    SDL_GetVersion(&v);
    printf("using SDL version %u.%u.%u\n", v.major, v.minor, v.patch);
    for (int i = 0; i < 20; i++) {
        SDL_AudioSpec req;
        req.freq = 44100;
        req.format = AUDIO_S16;
        req.samples = 4096;
        req.channels = 1;
        req.callback = NULL;

        SDL_Init(SDL_INIT_AUDIO);
        SDL_AudioDeviceID device_id = SDL_OpenAudioDevice(NULL, 0, &req, NULL, 0);

        printf("attempt %i\n", i);
        SDL_PauseAudioDevice(device_id, 0);
        SDL_Delay(256);

        SDL_Quit();
    }
}
```

Running this will produce output along the lines of 
```
using SDL version 2.30.0
attempt 0
attempt 1
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
attempt 2
attempt 3
attempt 4
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
attempt 5
attempt 6
attempt 7
attempt 8
attempt 9
attempt 10
attempt 11
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
attempt 12
attempt 13
attempt 14
attempt 15
attempt 16
attempt 17
attempt 18
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
ALSA lib pcm.c:8675:(snd_pcm_recover) underrun occurred
attempt 19
```

In my source program the underruns cause the output to sound clicky, here I demonstrated the behavior still occurs even when only silence is being written, though in this snippet it does not cause any clicking in the audio output.

This is being run on alpine linux 3.18.6, kernel 6.1.75-0-lts

**Comment by ericoporto** [comment]
@koltenpearson hey, can you try with the current release-2.30.x branch and see if it happens with it too?

**Comment by koltenpearson** [comment]
I tried on the release-2.30.x branch, and can confirm the issue is still
present

**Comment by koltenpearson** [comment]
I tried it on the tag release-2.0.0 and I can not replicate it there

**Comment by koltenpearson** [comment]
So it seems like the bug is not present in release-2.0.3 and is present in release 2.0.4

I did git bisect and got 0e45984fa036fff1912c62fee3ed4f5481f3c324, but I can't seem to see any parent commits of that commit so I am not sure it worked correctly.

**Comment by icculus** [comment]
> I did git bisect and got https://github.com/libsdl-org/SDL/commit/0e45984fa036fff1912c62fee3ed4f5481f3c324, but I can't seem to see any parent commits of that commit so I am not sure it worked correctly.

GitHub mangled our revision history when we imported from Mercurial (see #5220), so if you go far enough back, you always land in this commit, unfortunately.

**Comment by slouken** [comment]
https://github.com/spurious/SDL-mirror has full history, so you might be able to do a git bisect from there.

**Comment by koltenpearson** [comment]
okay from that repo I ran a git bisect and got a454256d88e413a4898b5005472625f755abfad4

**Comment by icculus** [comment]
Okay, if that's the culprit, it's been broken since we introduced SDL_QueueAudio. Notably, this also means it couldn't be broken on release-2.0.0 through release-2.0.3 which did not have this API yet. Maybe those were testing against the wrong build by accident?

I'm going to assign myself and see if this reproduces here. It's possible it has nothing to do with SDL_QueueAudio and ALSA is just being contrary in general, but I'll see what I can figure out over here.

**Comment by koltenpearson** [comment]
yeah I guess my example doesn't actually call SDL_QueueAudio, and so compiles through the api change. If before the SDL_QueueAudio is introduced, silence was not written to the buffer when the device was unpaused with no callback then it wouldn't be a matter of breaking anything, I agree.
